src/test/java/com/google/cloud/spanner/pgadapter/JdbcSimpleModeTest.java

```diff
@@ -0,0 +1,361 @@
+        // execute returns false if the result is an update count or no result.
+        assertFalse(statement.execute(String.format("%s; %s;", UPDATE_STATEMENT, SELECT2)));
+        assertEquals(2, statement.getUpdateCount());
+
+        assertTrue(statement.getMoreResults());
+        try (ResultSet resultSet = statement.getResultSet()) {
+          assertTrue(resultSet.next());
+          assertEquals(2L, resultSet.getLong(1));
+          assertFalse(resultSet.next());
+        }
+
+        // getMoreResults() should now return false. We should also check getUpdateCount() as that
+        // method should return -1 to indicate that there is also no update count available.
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+      }
+    }
+
+    // The server will also receive a 'SELECT 1' statement from PgAdapter, as PgAdapter calls
+    // connection.isAlive() before using it.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(3, requests.size());
+    assertEquals(SELECT1.getSql(), requests.get(0).getSql());
+    assertEquals(UPDATE_STATEMENT.getSql(), requests.get(1).getSql());
+    assertEquals(SELECT2.getSql(), requests.get(2).getSql());
+  }
+
+  @Test
+  public void testBeginAndDml() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(createUrl())) {
+      try (java.sql.Statement statement = connection.createStatement()) {
+        // batch a BEGIN [TRANSACTION] statement together with an update statement.
+        assertFalse(statement.execute(String.format("BEGIN; %s;", UPDATE_STATEMENT)));
+        assertEquals(0, statement.getUpdateCount());
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(2, statement.getUpdateCount());
+
+        // getMoreResults() should now return false. We should also check getUpdateCount() as that
+        // method should return -1 to indicate that there is also no update count available.
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+
+        // Execute an insert statement as part of the transaction.
+        assertFalse(statement.execute(INSERT_STATEMENT.getSql()));
+        assertEquals(1, statement.getUpdateCount());
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+
+        statement.execute("COMMIT");
+      }
+    }
+
+    // The server will also receive a 'SELECT 1' statement from PgAdapter, as PgAdapter calls
+    // connection.isAlive() before using it. This should happen before the transaction is started.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(3, requests.size());
+    assertEquals(SELECT1.getSql(), requests.get(0).getSql());
+    assertTrue(requests.get(0).getTransaction().hasSingleUse());
```

Comment:
What is this checking?

Reply:
This checks what (type of) transaction the statement uses. `hasSingleUse()` returns true if the request uses a single-use transaction. This means that the statement 'has its own transaction', so kind of an auto-commit like feature (in this case it is a read-only single use transaction).

This specific `SELECT 1` statement comes from the way PGAdapter used to open connections to Cloud Spanner. PGAdapter would first open a connection and then execute a `SELECT 1` statement to verify that the connection is valid. This has been removed since this code was written, as it is inefficient to do so for every connection that we open. Instead, the current implementation of PGAdapter relies on the automatic dialect detection in the client library to check that a connection is valid. This happens automatically in the background, and only the first time a connection to a specific database is opened.

The reason that this select statement should use a single-use read-only transaction is that it is not part of the transaction that is initiated by the batch (it is not at all related to the batch).